### ISSUE TITLE
Implement yr_scanner_scan_mem_blocks()-based io.Reader scanning interface?

### ISSUE THREAD

**Issue by jbremer** [hillu/go-yara#54]
As per the title, would it be possible to provide a scanning interface that accepts an `io.Reader`. The implementation could be based on a streaming interface by implementing the `YR_MEMORY_BLOCK_ITERATOR*` parameter for `yr_scanner_scan_mem_block()`?

With an `io.Reader` it'd be much easier to integrate `go-yara` in a clean and Go-like manner. Unless I missed something of course - happy to hear about alternatives (currently I believe there's either `ScanMem()`, which takes a singular `[]byte`, `ScanFile`, which takes a filename, and `ScanFileDescriptor`, which takes a fd) :-)

**Comment by hillu** [comment]
I believe that this should be doable. and would be quite useful Thank you for the suggestion.

**Comment by hillu** [comment]
@jbremer Please have a look at branch `scan-mem-blocks` which I just pushed.

**Comment by jbremer** [comment]
Very interesting, thank you for your work @hillu :-)

Built the following PoC, seems to work very well.

```go
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"os"

	"github.com/hillu/go-yara"
)

type Scan struct {
	offset int
	buf    []byte
	length int
	r      io.Reader
}

func (s *Scan) copy(buf []byte) {
	copy(buf, s.buf[:s.length])
}

func (s *Scan) First() *yara.MemoryBlock {
	return s.Next()
}

func (s *Scan) Next() *yara.MemoryBlock {
	n, err := s.r.Read(s.buf)
	if err != nil {
		return nil
	}
	s.offset += n
	s.length = n
	return &yara.MemoryBlock{
		Base:      uint64(s.offset - n),
		Size:      uint64(n),
		FetchData: s.copy,
	}
}

func scan(r *yara.Rules, f io.Reader) (yara.MatchRules, error) {
	var s Scan
	s.r = f
	s.buf = make([]byte, 0x10000)
	var mrs yara.MatchRules
	err := r.ScanMemBlocksWithCallback(&s, 0, 0, &mrs)
	return mrs, err
}

func main() {
	flag.Parse()
	f, err := os.Open(flag.Arg(0))
	if err != nil {
		log.Fatalln("error:", err)
	}
	defer f.Close()

	rs := yara.MustCompile(`
rule A {
	strings:
		$a = "abc"
		$b = "def"
	condition:
		all of them
}
	`, nil)

	mrs, err := scan(rs, f)
	if err != nil {
		fmt.Println("err:", err)
	}
	for _, rule := range mrs {
		fmt.Println("match:", rule)
	}
}
```

**Comment by hillu** [comment]
@jbremer For the general case, you'll find that you need an io.Reader that can be reset to the start of the file, otherwise simple rules such as ```rule B { condition: uint32(0) == 0 }``` won't work because YARA will read from the iterator again as it evaluates the conditions. Replacing the Reader in your example with a ReadSeeker and modifying the First method accordingly was enough for my test.

In any case, if you like the general interface and think that the functionality is solid, I'll be happy to merge it.

**Comment by jbremer** [comment]
Thanks. Yes, in that case a `ReadSeeker` should definitely do the job.
Excellent! Yeah, this interface is looking good to me. Please feel free to merge it :-)

**Comment by jbremer** [comment]
Small update, sample code implementing what you described:

```go
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"os"

	"github.com/hillu/go-yara"
)

type Scan struct {
	offset int
	buf    []byte
	length int
	r      io.ReadSeeker
}

func (s *Scan) copy(buf []byte) {
	copy(buf, s.buf[:s.length])
}

func (s *Scan) First() *yara.MemoryBlock {
	s.offset = 0
	s.r.Seek(0, io.SeekStart)
	return s.Next()
}

func (s *Scan) Next() *yara.MemoryBlock {
	n, err := s.r.Read(s.buf)
	if err != nil {
		return nil
	}
	s.offset += n
	s.length = n
	return &yara.MemoryBlock{
		Base:      uint64(s.offset - n),
		Size:      uint64(n),
		FetchData: s.copy,
	}
}

func scan(r *yara.Rules, f io.ReadSeeker) (yara.MatchRules, error) {
	var s Scan
	s.r = f
	s.buf = make([]byte, 0x10000)
	var mrs yara.MatchRules
	err := r.ScanMemBlocksWithCallback(&s, 0, 0, &mrs)
	return mrs, err
}

func main() {
	flag.Parse()
	f, err := os.Open(flag.Arg(0))
	if err != nil {
		log.Fatalln("error:", err)
	}
	defer f.Close()

	rs := yara.MustCompile(`
rule A {
	strings:
		$s1 = "abc"
		$s2 = "def"
	condition:
		uint8(0) == 0x7f and all of ($s*)
}
	`, nil)

	mrs, err := scan(rs, f)
	if err != nil {
		fmt.Println("err:", err)
	}
	for _, rule := range mrs {
		fmt.Println("match:", rule)
	}
}
```

**Comment by jbremer** [comment]
Hi @hillu, would it be possible to squeeze in one improvement still: namely that of being able to pass a `[]byte` slice to the `MemoryBlockIterator` struct such that the user can 1) define the buffer size 2) the additional `copy(...)` and malloc/realloc isn't necessary, essentially making it zero-copy.

**Comment by jbremer** [comment]
Hmm, changing the last part of the PoC to:

**Comment by hillu** [comment]
I had also noticed the base / offset issue, but had not looked further into it.

I'll add (*Match).Base(), exposing the `base` field from `YR_MATCH`.

**Comment by hillu** [comment]
@jbremer Unfortunately, we can't do without the `malloc`-backed buffer because the Go runtime will panic if we return a pointer to Go-allocated memory to C land. That being said, we can do with just one copy if we stick to the `YR_MEMROY_BLOCK_ITERATOR` logic and only do seek operations in the `First()` and `Next()` methods.

I have opened a pull request containing some example/test code demonstrating this. Let's use that PR to discuss any remaining issues.

**Comment by hillu** [comment]
Since I have merged #62, this can be closed. Thanks for the constructive discussion. :-)
